proto/kuksa/val/v2/val.proto

```diff
@@ -0,0 +1,233 @@
+
+  // Publish a signal value. Used for low frequency signals (e.g. attributes).
+  //
+  // Returns (GRPC error code):
+  //   NOT_FOUND if any of the signals are non-existant.
+  //   PERMISSION_DENIED
+  //       - if access is denied for any of the signals.
+  //       - if the signal is already provided by another provider.
+  //   INVALID_ARGUMENT
+  //       - if the data type used in the request does not match
+  //            the data type of the addressed signal
+  //       - if the published value is not accepted,
+  //            e.g. if sending an unsupported enum value
+  rpc PublishValue(PublishValueRequest) returns (PublishValueResponse);
+
+  // Open a stream used to provide actuation and/or publishing values using
+  // a streaming interface. Used to provide actuators and to enable high frequency
+  // updates of values.
+  //
+  // The open stream is used for request / response type communication between the
+  // provider and server (where the initiator of a request can vary).
+  // Errors are communicated as messages in the stream.
+  rpc OpenProviderStream(stream OpenProviderStreamRequest) returns (stream OpenProviderStreamResponse);
+
+  // Get server information
+  rpc GetServerInfo(GetServerInfoRequest) returns (GetServerInfoResponse);
+}
+
+message GetValueRequest {
+  SignalID signal_id = 1;
+}
+
+message GetValueResponse {
+  Datapoint data_point = 1;
+}
+
+message GetValuesRequest {
+  repeated SignalID signal_ids = 1;
+}
+
+message GetValuesResponse {
+  repeated Datapoint datapoints = 1;
+}
+
+message ListValuesRequest {
+  repeated SignalID signal_ids = 1;
+}
+
+message ListValuesResponse {
+  repeated Datapoint datapoints = 1;
+}
+
+message SubscribeRequest {
+  repeated SignalID signal_ids = 1;
+}
+
+message SubscribeResponse {
+  map<string, Datapoint> entries = 1;
+}
+
+message ActuateRequest {
+  SignalID signal_id = 1;
+  Value value        = 2;
+}
+
+message ActuateResponse {
+}
+
+message BatchActuateRequest {
+  repeated ActuateRequest actuate_requests = 1;
+}
+
+message BatchActuateResponse {
+}
+
+message ListMetadataRequest {
+  string root   = 1;
+  string filter = 2;
+}
+
+message ListMetadataResponse {
+  repeated Metadata metadata = 1;
+}
+
+message PublishValueRequest {
+  SignalID signal_id   = 1;
+  Datapoint data_point = 2;
+}
+
+message PublishValueResponse {
+  Error error = 1;
+}
+
+message PublishValuesRequest {
+  int32 request_id                  = 1; /// Unique request id for the stream that can be used to identify the response.
+  map<string, Datapoint> datapoints = 2;
+}
+
+message PublishValuesResponse {
+  int32 request_id          = 1;
+  map<string, Error> status = 2;
+}
+
+message ProvidedActuation {
+  repeated SignalID actuator_identifiers = 1;
+}
+
+message ProvideActuatorResponse {
+}
+
+message BatchActuateStreamRequest {
+  repeated ActuateRequest actuate_requests = 1;
+}
+
+message BatchActuateStreamResponse {
+}
+
+message OpenProviderStreamRequest {
```

Comment:
What is the rational of cramping multiple concerns into a single function call or stream? Is the idea to "save" on the number of open TCP connections?

To me this sounds like there could be problems on both side of the stream of correlating the different operations handled over the same stream. 
That is why it would feel more intuitive for me to have PublishValues as Stream and a ProvideActuation function here.

Reply:
But these calls most likely would be transmitted over the same gRCP channel anyway when they originate from the same client, right?